### ISSUE TITLE
Remove/reduce `format!` use in  `Interface::introspect_to_writer` impl in `interface`

### ISSUE THREAD

**Issue by zeenix** [z-galaxy/zbus#828]
This is adding to the binary size (#304).

The problem is number of `format!` calls and `format` macro generating a lot of code. Possible solutions:

* Convert the macros to create string literals instead. The issue is that we need some parameters that we don't have access to in the proc macro: indentation level and `Type::signature()` so it will not be as easy as simply moving the `format!` calls to the macro side.
* We could look into using `ufmt` crate but it seems to be not well-maintained and hasn't seen a release for 2 years. Perhaps some other crate?

**Comment by HookedBehemoth** [comment]
Could you elaborate on the issue with doing this at compile time some more? If it's all done then, indentation can be inferred there. And in my eyes, the signatures could also be obtained unless they are inferred.

**Comment by zeenix** [comment]
> Could you elaborate on the issue with doing this at compile time some more? If it's all done then, indentation can be inferred there.

You're right that the `level` argument of [`Interface::introspect_to_writer`]https://docs.rs/zbus/latest/zbus/object_server/trait.Interface.html#tymethod.introspect_to_writer can be changed to a const generic and we can do so w/o bumping the major version since `Interface` is declared to be unstable API.

However, the `indent` argument is bumped internally for each level here. Perhaps, the inner indentation level should be something the `Interface` implementation (i-e the macro) should handle and caller of `introspect_to_writer` should just pass the main indentation level. :thinking: 

> And in my eyes, the signatures could also be obtained unless they are inferred.

I only wish it was that easy. :cry: Currently in rust you can't concatenate strings at compile time unless they're themselves literals (trait constants can't work). I do have a plan now to make most signature handling (especially during ser/de) constant though, through parsed signatures (#882). I'm planning on working on this during the weekend/next week.

**Comment by zeenix** [comment]
So I went ahead and removed most format! calls but it doesn't save a hell lot in `busd`. The real issues are:

* The template strings being part of the binary.
* The number of calls, since we make separate calls to `writeln!` (or `fmt::Write::push_str` in my branch) for each argument of every method and signal.

 I am thinking of allowing interfaces to provide file paths to load introspection from. The will give users a way to completely eliminate both of the above (well not all the code but it'll be considerably less code to load a string from a file).

In the meantime, there #1136 that already helps a bit.

**Comment by zeenix** [comment]
🤖 ChatGPT on behalf of zeenix: The requested reduction is present on main: [introspection code generation](https://github.com/z-galaxy/zbus/blob/33dc2a77d4088232430269a645e1830405638b1b/zbus_macros/src/iface.rs#L1190-L1245) constructs the format strings inside the proc macro and emits `writeln!` calls using those literals, rather than allocating intermediate `format!` strings at runtime. Closing this optimization request; this does not claim all remaining introspection formatting overhead is eliminated.
